### PR TITLE
[MDB Ignore] DNA Vaults power check

### LINKED ISSUE #32373: DNA vaults can be used infinitely
As the title says you can infinitely use DNA vaults, though not in the way of getting infinite selections of powers just by building the thing. If you build another DNA vault after building and managing to fill up the first one and getting your free power you can build another one and get another free power and you can do this as many times as you like. This could be a feature, but you can do the same thing by building a DNA vault, completing it, and then destroying it by hitting it. the parts will drop and you'll be able to rebuild the entire thing and keep your upgrade and get another upgrade to go along with your first one, and you can do this as many times as you like so long as you have the patience.

### PR DESCRIPTION
**Pull request by Mooshimi**
## About The Pull Request

Fixes: #32373
by adding several checks for having received vault upgrades already

## Why It's Good For The Game

Eliminates an exploit involving building many DNA Vaults after the goal is complete and getting all of the powers.

## Changelog

:cl:
fix: Separate DNA Vaults will now check if you already have a power from one before giving a power to you.
/:cl:

### CODE REVIEW

**Comment by bluetshirtguy** [comment]
Wowie, less of a reason to build these things.

**Comment by Mooshimi** [comment]
> Wowie, less of a reason to build these things.

This removes an exploit with building multiple of them after completing the objective....

File: code/modules/station_goals/dna_vault.dm [hunk 1]
```diff
@@ -187,7 +187,7 @@
 		return ..()
 
 /obj/machinery/dna_vault/proc/upgrade(mob/living/carbon/human/H,upgrade_type)
-	if(!(upgrade_type in power_lottery[H]))
+	if(!(upgrade_type in power_lottery[H])||(HAS_TRAIT(H, TRAIT_VIRUSIMMUNE))||(HAS_TRAIT(H, TRAIT_NOBREATH))||(HAS_TRAIT(H, TRAIT_RESISTHEAT))||(HAS_TRAIT(H, TRAIT_NOFIRE))||(HAS_TRAIT(H, TRAIT_PIERCEIMMUNE))||(H.next_move_modifier == 0.5)||(H.has_movespeed_modifier(/datum/movespeed_modifier/dna_vault_speedup))||(H.dna.species.armor==30)||(H.dna.species.stunmod==0.5))
```
Comment (Fikou) [inline code/modules/station_goals/dna_vault.dm:190]: this is a very bad way to check if the user has one of the upgrades. people can receive or change almost all of these things you check for from other sources and then theyll be excluded

Comment (Mooshimi) [reply code/modules/station_goals/dna_vault.dm:190]: suggest an alternative, it doesn't give a vault specific trait

**Comment by Mooshimi** [comment]
Tested, and the two power options don't show on the secondary vault after getting a power on the first one.

File: code/modules/station_goals/dna_vault.dm [hunk 2]
```diff
@@ -219,5 +219,6 @@
 		if(VAULT_QUICK)
 			to_chat(H, span_notice("Your arms move as fast as lightning."))
 			H.next_move_modifier = 0.5
+	ADD_TRAIT(H, TRAIT_USED_DNA_VAULT, "used_dna_vault")
```
Comment (JohnFulpWillard) [inline code/modules/station_goals/dna_vault.dm:222]: add a source trait please

Comment (JohnFulpWillard) [inline code/modules/station_goals/dna_vault.dm:222]: ADD_TRAIT() has 3 args, H is what we're adding the trait to, ``TRAIT_USED_DNA_VAULT`` is the trait we're giving, and ``"used_dna_vault"`` is the source trait, where the trait is coming from.

Trait stuff including the source trait should be defines, so like how you made a define for ``TRAIT_USED_DNA_VAULT``, you should make a new one called ``DNA_VAULT_TRAIT``.

Comment (Mooshimi) [reply code/modules/station_goals/dna_vault.dm]: thanks for clarification

Comment (Mooshimi) [reply code/modules/station_goals/dna_vault.dm]: you mean here? 
image

File: code/modules/station_goals/dna_vault.dm [hunk 3]
```diff
@@ -219,5 +219,6 @@
 		if(VAULT_QUICK)
 			to_chat(H, span_notice("Your arms move as fast as lightning."))
 			H.next_move_modifier = 0.5
+	ADD_TRAIT(H, TRAIT_USED_DNA_VAULT, "dna_vault")
```
Comment (JohnFulpWillard) [inline code/modules/station_goals/dna_vault.dm:222]:
```suggestion
	ADD_TRAIT(H, TRAIT_USED_DNA_VAULT, DNA_VAULT_TRAIT)
```
Just like the trait you're giving, you need to use the define instead.

File: code/modules/station_goals/dna_vault.dm [hunk 4]
```diff
@@ -219,5 +219,6 @@
 		if(VAULT_QUICK)
 			to_chat(H, span_notice("Your arms move as fast as lightning."))
 			H.next_move_modifier = 0.5
+	ADD_TRAIT(H, TRAIT_USED_DNA_VAULT, DNA_VAULT_TRAIT)
```
Comment (ShizCalev) [inline code/modules/station_goals/dna_vault.dm:222]: update the 6 above it with the define too please.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/modules/station_goals/dna_vault.dm (hunk 4 above)
Comment (JohnFulpWillard) [reply code/modules/station_goals/dna_vault.dm:222]: This is given regardless of what you take from the vault

**Pull request merged**